File: react-native-pytorch-core/ios/Image/ImageDataImage.swift
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import Foundation

public class ImageDataImage: IImage {
    private var imageData: ImageData
    private var pixelDensity: CGFloat

    init(imageData: ImageData, pixelDensity: CGFloat) {
        self.imageData = imageData
        self.pixelDensity = pixelDensity
    }

    public func getPixelDensity() -> CGFloat {
        return self.pixelDensity
    }

    public func getWidth() -> CGFloat {
        return CGFloat(self.imageData.bitmap.width)
    }

    public func getHeight() -> CGFloat {
        return CGFloat(self.imageData.bitmap.height)
    }

    public func getNaturalWidth() -> Int {
        return Int(self.imageData.bitmap.width)
    }

    public func getNaturalHeight() -> Int {
        return Int(self.imageData.bitmap.height)
    }

    public func scale(sx: CGFloat, sy: CGFloat) throws -> IImage {
        if let cgImage = self.getBitmap() {
            let bitmapImage = Image(image: cgImage)
            return try ImageUtils.scale(image: bitmapImage, sx: sx, sy: sy)
        }
        throw ImageError.scale
    }

    public func getBitmap() -> CGImage? {
        return self.imageData.bitmap
    }

    public func close() throws {
        // TODO(T94684939)
    }
}

File: react-native-pytorch-core/android/src/main/java/org/pytorch/rn/core/image/ImageDataImage.java
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

package org.pytorch.rn.core.image;

import android.graphics.Bitmap;
import android.media.Image;
import androidx.annotation.Nullable;
import org.pytorch.rn.core.canvas.ImageData;

public class ImageDataImage extends AbstractImage {

  private final ImageData mImageData;
  private final float mPixelDensity;

  public ImageDataImage(ImageData imageData, float pixelDensity) {
    mImageData = imageData;
    mPixelDensity = pixelDensity;
  }

  @Override
  public float getPixelDensity() {
    return mPixelDensity;
  }

  @Override
  public float getNaturalWidth() {
    return mImageData.getWidth();
  }

  @Override
  public float getNaturalHeight() {
    return mImageData.getHeight();
  }

  @Override
  public Bitmap getBitmap() {
    return mImageData.getBitmap();
  }

  @Override
  public void close() throws Exception {
    mImageData.getBitmap().recycle();
  }

  @Nullable
  @Override
  public Image getImage() {
    return null;
  }

  @Override
  public int getImageRotationDegrees() {
    return 0;
  }
}
